File: php/src/Resource/Tasks.php
<?php

declare(strict_types=1);

namespace Fivexer\SDK\Resource;

use Fivexer\SDK\Fivexer;
use Fivexer\SDK\Model\AssignTaskResult;
use Fivexer\SDK\Model\BulkTaskReport;
use Fivexer\SDK\Model\CreateTask;
use Fivexer\SDK\Model\TaskCheckReport;
use Fivexer\SDK\Model\TaskEscalation;
use Fivexer\SDK\Model\TaskList;
use Fivexer\SDK\Model\UnparkTask;
use Fivexer\SDK\Model\SuggestWorkers;
use Fivexer\SDK\Model\SuggestWorkersResult;
use Fivexer\SDK\Model\Task;
use Fivexer\SDK\Model\TaskAction;
use Fivexer\SDK\Model\TaskPage;
use Fivexer\SDK\Model\TaskPriority;

/**
 * Task lifecycle resource. Delegates to Fivexer::request() so all transport logic
 * (auth, retry, quota, errors) lives in one place.
 */
final class Tasks
{
    private const PARAM_STATUS = 'status';
    private const PARAM_CURSOR = 'cursor';
    private const PARAM_LIMIT = 'limit';

    private readonly TaskContexts $contextResource;
    private readonly TaskComments $commentsResource;
    private readonly TaskAttachments $attachmentsResource;
    private readonly TaskRecurring $recurringResource;

    public function __construct(private readonly Fivexer $client)
    {
        $this->contextResource = new TaskContexts($client);
        $this->commentsResource = new TaskComments($client);
        $this->attachmentsResource = new TaskAttachments($client);
        $this->recurringResource = new TaskRecurring($client);
    }

    /** The rich data stored against a task: title, description, context and references. */
    public function context(): TaskContexts
    {
        return $this->contextResource;
    }

    /** Comments on a task. */
    public function comments(): TaskComments
    {
        return $this->commentsResource;
    }

    /** Files attached to a task. */
    public function attachments(): TaskAttachments
    {
        return $this->attachmentsResource;
    }

    /** Standing templates that occurrences are cut from. */
    public function recurring(): TaskRecurring
    {
        return $this->recurringResource;
    }

    /** Create a task; the server queues it for matching. */
    public function create(CreateTask $input): Task
    {
        $data = $this->client->request('POST', '/tasks', $input->toArray()) ?? [];
        return Task::fromArray($data);
    }

    public function get(string $taskId): Task
    {
        return Task::fromArray($this->client->request('GET', '/tasks/' . \rawurlencode($taskId)) ?? []);
    }

    /**
     * @param string|null $status 'queued'|'pending'|'accepted'|'all'
     * @param string|null $cursor Pagination cursor from a previous page's nextCursor
     * @param int<1, max>|null $limit Page size
     */
    public function list(?string $status = null, ?string $cursor = null, ?int $limit = null): TaskPage
    {
        $data = $this->client->request('GET', '/tasks', null, [
            self::PARAM_STATUS => $status,
            self::PARAM_CURSOR => $cursor,
            self::PARAM_LIMIT => $limit,
        ]) ?? [];
        return TaskPage::fromArray($data);
    }

    /** Cancel a queued task. */
    public function cancel(string $taskId): void
    {
        $this->client->request('DELETE', '/tasks/' . \rawurlencode($taskId));
    }

    /**
     * Create many tasks in one call.
     *
     * Partial success is normal — a 200 does not mean every task was created. Read `failed` and
     * the per-entry results, whose `index` maps back to this list.
     *
     * @param list<CreateTask> $tasks
     */
    public function createMany(array $tasks): BulkTaskReport
    {
        return BulkTaskReport::fromArray($this->client->request('POST', '/tasks/bulk', [
            'tasks' => \array_map(static fn (CreateTask $t): array => $t->toArray(), $tasks),
        ]) ?? []);
    }

    /**
     * Dry-run a task: how many workers could take it, which of its tags nobody covers, and what
     * is wrong with its policies. Creates nothing and reserves nothing.
     */
    public function check(CreateTask $input): TaskCheckReport
    {
        return TaskCheckReport::fromArray($this->client->request('POST', '/tasks/check', $input->toArray()) ?? []);
    }

    /** Acknowledge an offer without starting work — stops the response clock only. */
    public function ack(string $taskId, string $workerId): TaskAction
    {
        return TaskAction::fromArray(
            $this->client->request('POST', '/tasks/' . \rawurlencode($taskId) . '/ack', [
                'workerId' => $workerId,
            ]) ?? []
        );
    }

    /**
     * Advance the escalation ladder now. `parked` comes back true when the ladder was already
     * exhausted, which takes the task out of matching.
     */
    public function escalate(string $taskId, ?string $workerId = null): TaskEscalation
    {
        $body = $workerId === null ? [] : ['workerId' => $workerId];
        return TaskEscalation::fromArray(
            $this->client->request('POST', '/tasks/' . \rawurlencode($taskId) . '/escalate', $body) ?? []
        );
    }

    /**
     * Tasks a policy gave up on — an exhausted escalation ladder, an SLA breach handled with
     * park, or a rejection budget run dry. Out of matching, but recoverable with unpark().
     */
    public function parked(): TaskList
    {
        return TaskList::fromArray($this->client->request('GET', '/tasks/parked') ?? []);
    }

    /**
     * Tasks held by a `schedule.notBefore`, soonest activation first — the only view of work
     * that has been booked but has not started.
     */
    public function scheduled(): TaskList
    {
        return TaskList::fromArray($this->client->request('GET', '/tasks/scheduled') ?? []);
    }

    /**
     * Return a parked task to the queue. Reset the clocks that parked it, or the next sweep may
     * park it straight back.
     */
    public function unpark(string $taskId, ?UnparkTask $options = null): TaskAction
    {
        return TaskAction::fromArray(
            $this->client->request(
                'POST',
                '/tasks/' . \rawurlencode($taskId) . '/unpark',
                $options === null ? [] : $options->toArray()
            ) ?? []
        );
    }

    public function accept(string $taskId, string $workerId): TaskAction
    {
        return $this->workerAction($taskId, 'accept', $workerId);
    }

    public function reject(string $taskId, string $workerId): TaskAction
    {
        return $this->workerAction($taskId, 'reject', $workerId);
    }

    /** @param array<string, mixed>|null $result */
    public function complete(string $taskId, string $workerId, ?array $result = null): TaskAction
    {
        $body = ['workerId' => $workerId];
        if ($result !== null) {
            $body['result'] = $result;
        }
        $data = $this->client->request('POST', '/tasks/' . \rawurlencode($taskId) . '/complete', $body) ?? [];
        return TaskAction::fromArray($data);
    }

    /** DRY helper: accept/reject share the exact same shape (workerId only). */
    private function workerAction(string $taskId, string $action, string $workerId): TaskAction
    {
        $data = $this->client->request(
            'POST',
            '/tasks/' . \rawurlencode($taskId) . '/' . $action,
            ['workerId' => $workerId],
        ) ?? [];
        return TaskAction::fromArray($data);
    }

    /**
     * Operator override: hand a task to a specific worker. Queued tasks go through the same
     * claim gate as organic matching; a pending task is transferred and its expiry clock
     * restarts. The result names the worker it was taken from, if any.
     *
     * @param bool $force Bypass the paused/backlog/veto/prior-rejection checks (worker
     *     existence is still enforced)
     */
    public function assign(string $taskId, string $workerId, bool $force = false): AssignTaskResult
    {
        $body = ['workerId' => $workerId];
        if ($force) {
            $body['force'] = true;
        }
        return AssignTaskResult::fromArray(
            $this->client->request('POST', '/tasks/' . \rawurlencode($taskId) . '/assign', $body) ?? []
        );
    }

    public function setPriority(string $taskId, float $priority): TaskPriority
    {
        return TaskPriority::fromArray(
            $this->client->request('PATCH', '/tasks/' . \rawurlencode($taskId), ['priority' => $priority]) ?? []
        );
    }

    /** Dry run: who *would* match these tags, without creating a task. */
    public function suggestWorkers(SuggestWorkers $input): SuggestWorkersResult
    {
        return SuggestWorkersResult::fromArray(
            $this->client->request('POST', '/tasks/suggest-workers', $input->toArray()) ?? []
        );
    }
}

File: java/src/main/java/io/fivexer/sdk/Tasks.java
package io.fivexer.sdk;

import io.fivexer.sdk.internal.Json;
import io.fivexer.sdk.model.BulkTaskReport;
import io.fivexer.sdk.model.TaskCheckReport;
import io.fivexer.sdk.model.TaskEscalation;
import io.fivexer.sdk.model.TaskList;
import io.fivexer.sdk.model.UnparkTask;
import io.fivexer.sdk.model.AssignTaskResult;
import io.fivexer.sdk.model.CreateTask;
import io.fivexer.sdk.model.SuggestWorkers;
import io.fivexer.sdk.model.SuggestWorkersResult;
import io.fivexer.sdk.model.Task;
import io.fivexer.sdk.model.TaskAction;
import io.fivexer.sdk.model.TaskPage;
import io.fivexer.sdk.model.TaskPriority;
import java.util.Map;

/** The task lifecycle, plus the operator overrides and the rich-data sub-resources. */
public final class Tasks {

    private final Fivexer client;
    private final TaskContexts contextResource;
    private final TaskComments commentsResource;
    private final TaskAttachments attachmentsResource;
    private final TaskRecurring recurringResource;

    Tasks(Fivexer client) {
        this.client = client;
        this.contextResource = new TaskContexts(client);
        this.commentsResource = new TaskComments(client);
        this.attachmentsResource = new TaskAttachments(client);
        this.recurringResource = new TaskRecurring(client);
    }

    /** The rich data stored against a task: title, description, context and references. */
    public TaskContexts context() { return contextResource; }

    /** Comments on a task. */
    public TaskComments comments() { return commentsResource; }

    /** Files attached to a task. */
    public TaskAttachments attachments() { return attachmentsResource; }

    /** Standing templates that occurrences are cut from. */
    public TaskRecurring recurring() { return recurringResource; }

    /** Enqueue a task. The response carries the id and {@code queued} status. */
    public Task create(CreateTask input) {
        return client.request("POST", "/tasks", input.toJson(), null, Task.class, false);
    }

    public Task get(String taskId) {
        return client.request("GET", "/tasks/" + Json.enc(taskId), null, null, Task.class, false);
    }

    public TaskPage list(ListTasksQuery query) {
        ListTasksQuery q = query == null ? new ListTasksQuery() : query;
        return client.request("GET", "/tasks", null, q.toQuery(), TaskPage.class, false);
    }

    public void cancel(String taskId) {
        client.request("DELETE", "/tasks/" + Json.enc(taskId), null, null, Void.class, true);
    }

    /**
     * Create many tasks in one call.
     *
     * <p>Partial success is normal — a 200 does not mean every task was created. Read {@code
     * failed} and the per-entry results, whose {@code index} maps back to this list.
     */
    public BulkTaskReport createMany(java.util.List<CreateTask> tasks) {
        com.google.gson.JsonArray array = new com.google.gson.JsonArray();
        for (CreateTask task : tasks) {
            array.add(com.google.gson.JsonParser.parseString(task.toJson()));
        }
        com.google.gson.JsonObject body = new com.google.gson.JsonObject();
        body.add("tasks", array);
        return client.request("POST", "/tasks/bulk", body.toString(), null,
                BulkTaskReport.class, false);
    }

    /**
     * Dry-run a task: how many workers could take it, which of its tags nobody covers, and what
     * is wrong with its policies. Creates nothing and reserves nothing.
     */
    public TaskCheckReport check(CreateTask input) {
        return client.request("POST", "/tasks/check", input.toJson(), null,
                TaskCheckReport.class, false);
    }

    /** Acknowledge an offer without starting work — stops the response clock only. */
    public TaskAction ack(String taskId, String workerId) {
        return client.request("POST", "/tasks/" + Json.enc(taskId) + "/ack",
                Json.workerAction(workerId, null), null, TaskAction.class, false);
    }

    /**
     * Advance the escalation ladder now. {@code parked} comes back true when the ladder was
     * already exhausted, which takes the task out of matching.
     */
    public TaskEscalation escalate(String taskId, String workerId) {
        String body = workerId == null ? "{}" : Json.object("workerId", workerId);
        return client.request("POST", "/tasks/" + Json.enc(taskId) + "/escalate", body, null,
                TaskEscalation.class, false);
    }

    public TaskEscalation escalate(String taskId) {
        return escalate(taskId, null);
    }

    /**
     * Tasks a policy gave up on — an exhausted escalation ladder, an SLA breach handled with
     * park, or a rejection budget run dry. Out of matching, but recoverable with {@link #unpark}.
     */
    public TaskList parked() {
        return client.request("GET", "/tasks/parked", null, null, TaskList.class, false);
    }

    /**
     * Tasks held by a {@code schedule.notBefore}, soonest activation first — the only view of
     * work that has been booked but has not started.
     */
    public TaskList scheduled() {
        return client.request("GET", "/tasks/scheduled", null, null, TaskList.class, false);
    }

    /**
     * Return a parked task to the queue. Reset the clocks that parked it, or the next sweep may
     * park it straight back.
     */
    public TaskAction unpark(String taskId, UnparkTask options) {
        String body = options == null ? "{}" : options.toJson();
        return client.request("POST", "/tasks/" + Json.enc(taskId) + "/unpark", body, null,
                TaskAction.class, false);
    }

    public TaskAction unpark(String taskId) {
        return unpark(taskId, null);
    }

    public TaskAction accept(String taskId, String workerId) {
        return client.request("POST", "/tasks/" + Json.enc(taskId) + "/accept",
                Json.workerAction(workerId, null), null, TaskAction.class, false);
    }

    /** Reject a task — it requeues for other eligible workers. */
    public TaskAction reject(String taskId, String workerId) {
        return client.request("POST", "/tasks/" + Json.enc(taskId) + "/reject",
                Json.workerAction(workerId, null), null, TaskAction.class, false);
    }

    public TaskAction complete(String taskId, String workerId, Map<String, Object> result) {
        return client.request("POST", "/tasks/" + Json.enc(taskId) + "/complete",
                Json.workerAction(workerId, result), null, TaskAction.class, false);
    }

    /**
     * Operator override: hand a task to a specific worker. Queued tasks go through the same
     * claim gate as organic matching; a pending task is transferred and its expiry clock
     * restarts. The result names the worker it was taken from, if any.
     */
    public AssignTaskResult assign(String taskId, String workerId) {
        return assign(taskId, workerId, false);
    }

    /**
     * @param force bypass the paused/backlog/veto/prior-rejection checks (worker existence is
     *     still enforced)
     */
    public AssignTaskResult assign(String taskId, String workerId, boolean force) {
        com.google.gson.JsonObject body = new com.google.gson.JsonObject();
        body.addProperty("workerId", workerId);
        if (force) {
            body.addProperty("force", true);
        }
        return client.request("POST", "/tasks/" + Json.enc(taskId) + "/assign",
                body.toString(), null, AssignTaskResult.class, false);
    }

    public TaskPriority setPriority(String taskId, double priority) {
        return client.request("PATCH", "/tasks/" + Json.enc(taskId),
                Json.object("priority", priority), null, TaskPriority.class, false);
    }

    /** Dry run: who <em>would</em> match these tags, without creating a task. */
    public SuggestWorkersResult suggestWorkers(SuggestWorkers input) {
        return client.request("POST", "/tasks/suggest-workers", input.toJson(), null,
                SuggestWorkersResult.class, false);
    }
}
